### PR TITLE
[KIP-951] Fetch RPC prework: upgrade to v15

### PR DESCRIPTION
**Pull request by mahajanadhitya**

### CODE REVIEW

File: src/rdkafka_fetcher.c
Comment (emasab) [inline src/rdkafka_fetcher.c:417]: remove `k` here and use the one at the top of the function

File: src/rdkafka_fetcher.c
Comment (emasab) [inline src/rdkafka_fetcher.c:600]: remove this additional line

File: src/rdkafka_topic.h
Comment (emasab) [inline src/rdkafka_topic.h:259]: This is the cause of the compilation error, remove the implementation and leave only the declaration here.
Also, remove the comment as it's defined in the implementation.

```suggestion
rd_kafka_topic_t *rd_kafka_topic_find_by_topic_id(rd_kafka_t *rk, rd_kafka_Uuid_t topic_id);
```

File: src/rdkafka_fetcher.c
Comment (emasab) [inline src/rdkafka_fetcher.c:408]: there was a for loop here.
Also `k `definition should go to the top of the function.

```suggestion
                                for (k = 0; k < AbortedTxnCnt; k++) {
                                        rd_kafka_buf_skip(rkbuf, (8 + 8));
                                        /* AbortedTransaction tags */
                                        rd_kafka_buf_skip_tags(rkbuf);
                                }
```

**Review by emasab — CHANGES_REQUESTED** [review]

File: src/rdkafka_topic.h
Comment (pranavrth) [inline src/rdkafka_topic.h:137]: use `rkt_topic_id` instead.

File: src/rdkafka_fetcher.c
Comment (pranavrth) [inline src/rdkafka_fetcher.c:887]: We can use the common function to write the toppar here?

`rd_kafka_buf_write_topic_partitions`

File: src/rdkafka_fetcher.c
Comment (pranavrth) [inline src/rdkafka_fetcher.c:653]: We should use common function `rd_kafka_buf_read_topic_partitions` to read this topic partition.

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: src/rdkafka_fetcher.c
Comment (pranavrth) [inline src/rdkafka_fetcher.c:793]: We should increase the buffer size

File: src/rdkafka_mock_int.h
Comment (pranavrth) [inline src/rdkafka_mock_int.h:439]: What does this mean? Better name?

**Review by pranavrth — COMMENTED** [review]
Some comments and discussion points.

**Comment by pranavrth** [comment]
Are we comparing the performance after these changes with the old one? Are we planning to do that for KIP-951 atleast?

File: src/rdkafka_buf.h
Comment (emasab) [reply src/rdkafka_buf.h:828]: It's in a later PR

File: src/rdkafka_fetcher.c
Comment (emasab) [reply src/rdkafka_fetcher.c:647]: AK 3.0.2 has Metadata v11 and Fetch v12. AK 3.1.0 has Metadata v12 and Fetch v13 so the topic id was enabled in the same version.

File: src/rdkafka_mock_int.h
Comment (emasab) [reply src/rdkafka_mock_int.h:439]: As the language doesn't have overloading it's often used in the repo to make the function ending with 0 the one with all the parameters and the implementation and other functions or defines call it. I understand that isn't not clear and would prefer naming it like `*_generic(` but would prefer to do the refactor all at once to avoid jeopardizing the conventions.

**Comment by emasab** [comment]
@pranavrth about performance I've run the performance test (0038) on versions from 1.9.2 to this branch, increasing totsize to 256 MiB to decrease variance. I've taken the best 3 of 4 runs. AK 3.7.0 in KRaft mode, Trivup defaults, local cluster.

The rates remained stable.

```
This branch:

Report #0: { "producer":  { "mb_per_sec": 82.67, "records_per_sec": 826746.44 }, "consumer": { "mb_per_sec": 22.29, "records_per_sec": 222914.94 } }
Report #0: { "producer":  { "mb_per_sec": 88.99, "records_per_sec": 889882.00 }, "consumer": { "mb_per_sec": 24.40, "records_per_sec": 244030.97 } }
Report #0: { "producer":  { "mb_per_sec": 98.61, "records_per_sec": 986124.81 }, "consumer": { "mb_per_sec": 24.87, "records_per_sec": 248658.94 } }

v2.4.0

Report #0: { "producer":  { "mb_per_sec": 91.76, "records_per_sec": 917647.25 }, "consumer": { "mb_per_sec": 20.66, "records_per_sec": 206640.33 } }
Report #0: { "producer":  { "mb_per_sec": 94.12, "records_per_sec": 941154.38 }, "consumer": { "mb_per_sec": 22.67, "records_per_sec": 226655.03 } }
Report #0: { "producer":  { "mb_per_sec": 87.32, "records_per_sec": 873185.56 }, "consumer": { "mb_per_sec": 22.44, "records_per_sec": 224371.58 } }

v2.3.0

Report #0: { "producer":  { "mb_per_sec": 107.63, "records_per_sec": 1076326.12 }, "consumer": { "mb_per_sec": 24.49, "records_per_sec": 244922.56 } }
Report #0: { "producer":  { "mb_per_sec": 119.94, "records_per_sec": 1199374.12 }, "consumer": { "mb_per_sec": 24.90, "records_per_sec": 248968.25 } }
Report #0: { "producer":  { "mb_per_sec": 93.76, "records_per_sec": 937584.81 }, "consumer": { "mb_per_sec": 24.36, "records_per_sec": 243566.44 } }

v2.2.0

Report #0: { "producer":  { "mb_per_sec": 114.31, "records_per_sec": 1143102.75 }, "consumer": { "mb_per_sec": 27.05, "records_per_sec": 270502.81 } }
Report #0: { "producer":  { "mb_per_sec": 123.92, "records_per_sec": 1239186.88 }, "consumer": { "mb_per_sec": 24.93, "records_per_sec": 249310.66 } }
Report #0: { "producer":  { "mb_per_sec": 111.29, "records_per_sec": 1112932.38 }, "consumer": { "mb_per_sec": 24.63, "records_per_sec": 246279.45 } }

v2.1.1

Report #0: { "producer":  { "mb_per_sec": 94.03, "records_per_sec": 940311.06 }, "consumer": { "mb_per_sec": 25.42, "records_per_sec": 254182.95 } }
Report #0: { "producer":  { "mb_per_sec": 117.41, "records_per_sec": 1174142.50 }, "consumer": { "mb_per_sec": 21.33, "records_per_sec": 213310.50 } }
Report #0: { "producer":  { "mb_per_sec": 112.83, "records_per_sec": 1128308.38 }, "consumer": { "mb_per_sec": 28.20, "records_per_sec": 282045.38 } }

v2.0.2

Report #0: { "producer":  { "mb_per_sec": 94.49, "records_per_sec": 944855.06 }, "consumer": { "mb_per_sec": 19.80, "records_per_sec": 198008.22 } }
Report #0: { "producer":  { "mb_per_sec": 125.44, "records_per_sec": 1254406.75 }, "consumer": { "mb_per_sec": 21.28, "records_per_sec": 212789.23 } }
Report #0: { "producer":  { "mb_per_sec": 112.99, "records_per_sec": 1129943.12 }, "consumer": { "mb_per_sec": 21.44, "records_per_sec": 214449.66 } }

v1.9.2

Report #0: { "producer":  { "mb_per_sec": 106.98, "records_per_sec": 1069808.12 }, "consumer": { "mb_per_sec": 21.44, "records_per_sec": 214372.36 } }
Report #0: { "producer":  { "mb_per_sec": 94.71, "records_per_sec": 947143.69 }, "consumer": { "mb_per_sec": 21.23, "records_per_sec": 212328.94 } }
Report #0: { "producer":  { "mb_per_sec": 110.96, "records_per_sec": 1109618.62 }, "consumer": { "mb_per_sec": 21.28, "records_per_sec": 212792.20 } }

```

File: src/rdkafka_fetcher.c
Comment (anchitj) [inline src/rdkafka_fetcher.c:671]: Add comment /* Topic Tags */

File: src/rdkafka_fetcher.c
Comment (anchitj) [inline src/rdkafka_fetcher.c:425]: Why did we remove this?

File: src/rdkafka_mock_handlers.c
Comment (anchitj) [inline src/rdkafka_mock_handlers.c:400]: /* Response: Partition tags */

File: src/rdkafka_mock_handlers.c
Comment (anchitj) [inline src/rdkafka_mock_handlers.c:443]: /* Response: Top level tags */

**Review by anchitj — COMMENTED** [review]
Overall LGTM! Some minor comments

**Review by anchitj — APPROVED** [review]

**Review by emasab — APPROVED** [review]
Approving after the changes, only @pranavrth approval is missing

**Review by pranavrth — COMMENTED** [review]
Why is our performance decreasing by each version? This branch has even more decreased performance.

File: src/rdkafka_mock_int.h
Comment (pranavrth) [reply src/rdkafka_mock_int.h:439]: I was talking about `tags_written` parameter. Can we name that better?

**Review by pranavrth — APPROVED** [review]

**Pull request merged**
